src/libraries/System.Resources.Extensions/src/System/Resources/Extensions/DeserializingResourceReader.cs

```diff
@@ -37,18 +35,40 @@ private bool ValidateReaderType(string readerType)
             return false;
         }
 
-// Issue https://github.com/dotnet/runtime/issues/39292 tracks finding an alternative to BinaryFormatter
-#pragma warning disable SYSLIB0011
         private object ReadBinaryFormattedObject()
         {
-            _formatter ??= new BinaryFormatter()
+            if (UseBinaryFormatter)
             {
-                Binder = new UndoTruncatedTypeNameSerializationBinder()
-            };
+                return DeserializeUsingBinaryFormatter(_store.BaseStream);
+            }
 
-            return _formatter.Deserialize(_store.BaseStream);
-        }
+            long position = _store.BaseStream.CanSeek ? _store.BaseStream.Position : -1;
+
+            try
+            {
+                BinaryFormattedObject binaryFormattedObject = new(_store.BaseStream);
+
+                return binaryFormattedObject.Deserialize();
+            }
+            catch (NotSupportedException) when (position >= 0)
+            {
+                _store.BaseStream.Position = position;
+
+                return DeserializeUsingBinaryFormatter(_store.BaseStream);
```

Comment:
What's the point of this fallback? If the app needs the full compatible binary formatter, it should better use it explicitly.

Reply:
> What's the point of this fallback? If the app needs the full compatible binary formatter, it should better use it explicitly.

The main reason are... inputs generated by some MS tooling that were generating invalid generic type names for years.

Example: `Dictionary<string, string>` has `KeyValuePair<string, string>[]` inside, but the forementioned tooling was generating:
- type name: ``System.Collections.Generic.Dictionary`2[[System.String``
- library name: `mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]`

I've tried to handle that mangling: https://github.com/dotnet/runtime/pull/102379/commits/6354a91c6a12a24fb58933bf6cda936bcef2852e

But I've find out that it does not affect only records where both type and assembly name are provided (I can join them if I really want to), but also field information, where part of the information is simply missing. BF made it work because it simply does not care whether the field information matches the record type information that is provided later in the payload. We were also using a binder that was doing some hacks like string concat to get it to work:
https://github.com/dotnet/runtime/blob/07782857083c9ae1fd98aa37bdda1f7bd1bb0ab1/src/libraries/System.Resources.Extensions/src/System/Resources/Extensions/DeserializingResourceReader.cs#L53-L72

PayloadReader treats such inputs as invalid and rejects them by design.

We have some tests that ensure that this keeps working. If I remove the fallback, I need to disable them and enable only in the compat test project. I am fine with that, what I don't know is how popular the mangling is in the real-life.

cc @ericstj @GrabYourPitchforks

Reply:
>  inputs generated by some MS tooling that were generating invalid generic type names for years.

It's not just "some tooling" it's every resource that's read from ResX that has a name with a comma somewhere in the type portion.  https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/Resources/ResXDataNode.cs,839

I thought you were going to make the PayloadReader support this scenario?  Can't the reader have some sort of binder that reverses the mangling?

Reply:
I don't think we want the fallback, other than the AppContext fallback.

Either we have confidence in the new deserializer, or we shouldn't ship it.

Reply:
> It's not just "some tooling" 

I used term "some tooling" because I forgot the exact name, I did not mean to depreciate the tooling itself.

I've implemented this support and I am about to push the change. Now all I need is a nice name for the config flag ;)